RecoTracker/DeDx/plugins/SiPixelTrackProbQXYProducer.cc

```diff
@@ -0,0 +1,187 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+
+//#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/StripGeomDetUnit.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "TrackingTools/PatternTools/interface/TrajTrackAssociation.h"
+#include "TrackingTools/TransientTrackingRecHit/interface/TransientTrackingRecHit.h"
+#include "DataFormats/TrackerRecHit2D/interface/SiPixelRecHit.h"
+#include "DataFormats/TrackReco/interface/SiPixelTrackProbQXY.h"
+
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "Geometry/Records/interface/TrackerTopologyRcd.h"
+
+//
+// class declaration
+//
+
+class SiPixelTrackProbQXYProducer : public edm::stream::EDProducer<> {
+public:
+  explicit SiPixelTrackProbQXYProducer(const edm::ParameterSet&);
+  ~SiPixelTrackProbQXYProducer() override = default;
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  int factorial(int);
+
+  // ----------member data ---------------------------
+  edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> tTopoToken_;
+  edm::EDGetTokenT<reco::TrackCollection> trackToken_;
+};
+
+using namespace reco;
+using namespace std;
+using namespace edm;
+
+SiPixelTrackProbQXYProducer::SiPixelTrackProbQXYProducer(const edm::ParameterSet& iConfig) : tTopoToken_(esConsumes()) {
+  produces<reco::SiPixelTrackProbQXYCollection>();
+  produces<reco::SiPixelTrackProbQXYAssociation>();
+  trackToken_ = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("tracks"));
+}
+
+void SiPixelTrackProbQXYProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<reco::TrackCollection> trackCollectionHandle;
+  iEvent.getByToken(trackToken_, trackCollectionHandle);
+  const TrackCollection& trackCollection(*trackCollectionHandle.product());
+  float probQonTrack = 0.0;
+  float probXYonTrack = 0.0;
+  float probQonTrackNoL1 = 0.0;
+  float probXYonTrackNoL1 = 0.0;
+  // creates the output collection
+  auto resultSiPixelTrackProbQXYColl = std::make_unique<reco::SiPixelTrackProbQXYCollection>();
+  std::vector<int> indices;
+
+  // Loop through the tracks
+  for (unsigned int j = 0; j < trackCollection.size(); j++) {
+    const reco::Track& track = trackCollection[j];
+
+    int numRecHits = 0;
+    int numRecHitsNoL1 = 0;
+    float probQonTrackWMulti = 1;
+    float probXYonTrackWMulti = 1;
+    float probQonTrackWMultiNoL1 = 1;
+    float probXYonTrackWMultiNoL1 = 1;
+
+    // Loop through the rechits on the given track
+    auto hb = track.recHitsBegin();
+    for (unsigned int h = 0; h < track.recHitsSize(); h++) {
```

Comment:
```suggestion
    for(auto const& hit : track.recHits()) 
```
and changing the continuation accordingly.